File: solutions/138.copy-list-with-random-pointer.js
/*
 * @lc app=leetcode id=138 lang=javascript
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function (head, m = new WeakMap()) {
  if (!head) return null;
  if (m.has(head)) return m.get(head);

  const clone = new Node(head.val);
  m.set(head, clone);
  clone.next = copyRandomList(head.next, m);
  clone.random = copyRandomList(head.random, m);
  return clone;
};
// @lc code=end

File: solutions/138.copy-list-with-random-pointer.rb
# Definition for Node.
# class Node
#     attr_accessor :val, :next, :random
#     def initialize(val = 0)
#         @val = val
#		  @next = nil
#		  @random = nil
#     end
# end

# @param {Node} node
# @return {Node}
def copyRandomList(head, memo={})
  return if head.nil?
  return memo[head] if memo.key? head

  Node.new(head.val).tap {
    memo[head] = _1
    _1.next = copyRandomList(head.next, memo)
    _1.random = copyRandomList(head.random, memo)
  }
end
